sys/mips/cheri/support.S

```diff
@@ -0,0 +1,831 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+ *	Contains capability-enabled assembly language support routines.
+ */
+
+#include "opt_ddb.h"
+#include <sys/errno.h>
+#include <machine/asm.h>
+#include <machine/cpu.h>
+#include <machine/regnum.h>
+#include <machine/cpuregs.h>
+#include <machine/pcb.h>
+#include <machine/cheriasm.h>
+
+#include <cheri/cherireg.h>
+
+#include "assym.inc"
+
+	.set	cheri_sysregs_accessible
+	.set	noreorder
+
+	.text
+
+/*
+ * pure capability ABI-only versions of the support routines
+ */
+
+/*
+ * Copy a null terminated string from the user address space into
+ * the kernel address space.
+ *
+ * copyinstr(caddr_t fromaddr, caddr_t toaddr, u_int maxlength, uint *lencopied)
+ * accepts:
+ * c3 - source address
+ * c4 - destination address
+ * c5 - pointer length copied
+ * a0 - maximum length
+ */
+NESTED(copyinstr, CALLFRAME_SIZ, ra)
+	cincoffset	CHERI_REG_STC, CHERI_REG_STC, -CALLFRAME_SIZ
+	csc		$c17, zero, CALLFRAME_CRA(CHERI_REG_STC)
+
+	cgetbase	t0, $c3 # make sure source is in userspace
+	bge	t0, zero, 1f
+	nop
+
+	CAPTABLE_LOAD($c12, copyerr)
+	cjr	$c12
+	nop
+1:
+	GET_CPU_PCPU($c1)	# clobbers at
+	clc	$c1, zero, PC_CURPCB($c1)
+	CAPTABLE_LOAD($c2, copyerr)	# prepare the onfault jump point
+	CAPCALL_LOAD($c12, copystr_c)
+	cjalr	$c12, $c17
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
+
+	GET_CPU_PCPU($c1)
+	clc	$c1, zero, PC_CURPCB($c1)
+	CHERI_NULL($c2)
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
+
+	clc		$c17, zero, CALLFRAME_CRA(CHERI_REG_STC)
+	/* Propagate v0 return value from copyinstr */
+	cjr	$c17
+	cincoffset	CHERI_REG_STC, CHERI_REG_STC, CALLFRAME_SIZ
+END(copyinstr)
+
+/*
+ * Copy a null terminated string from the kernel address space into
+ * the user address space.
+ *
+ * copyoutstr(caddr_t fromaddr, caddr_t toaddr, u_int maxlength, u_int *lencopied)
+ * accepts:
+ * c3 - source address
+ * c4 - destination address
+ * c5 - pointer length copied
+ * a0 - maximum length
+ */
+NESTED(copyoutstr, CALLFRAME_SIZ, ra)
+	cincoffset	CHERI_REG_STC, CHERI_REG_STC, -CALLFRAME_SIZ
+	csc		$c17, zero, CALLFRAME_CRA(CHERI_REG_STC)
+
+	cgetbase	t0, $c4 # make sure destination is in userspace
+	bge	t0, zero, 1f
+	nop
+
+	CAPTABLE_LOAD($c12, copyerr)
+	cjr	$c12
+	nop
+1:
+	GET_CPU_PCPU($c1)	# clobbers at
+	clc	$c1, zero, PC_CURPCB($c1)
+	CAPTABLE_LOAD($c2, copyerr)	# prepare the onfault jump point
+	CAPCALL_LOAD($c12, copystr_c)
+	cjalr	$c12, $c17
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
+
+	GET_CPU_PCPU($c1)
+	clc	$c1, zero, PC_CURPCB($c1)
+	CHERI_NULL($c2)
+	csc	$c2, zero, U_PCB_ONFAULT($c1)
```

Comment:
```suggestion
	csc	$cnull, zero, U_PCB_ONFAULT($c1)
```
Probably not much point wasting time improving MIPS assembly, but I just saw this macro seems unnecessary.

Reply:
Yes, I attempted to remove all uses of `CHERI_NULL`, I think it might have been in another PR.